go-controller/pkg/ovn/logical_switch_cache.go

```diff
@@ -0,0 +1,165 @@
+		klog.Warningf("Node %q logical switch already in cache with subnet %s; replacing with %s", nodeName,
+			util.JoinIPNets(existing.hostSubnets, ","), util.JoinIPNets(hostSubnets, ","))
+	}
+	var ipams []ipam.Interface
+	for _, subnet := range hostSubnets {
+		ipam := cache.ipamFunc(subnet)
+		if ipam == nil {
+			klog.Warningf("IPAM for subnet %s was not initialized for node %q", subnet, nodeName)
+			continue
+		}
+		// reserve IPs for gw, management port, hybrid overlay etc.
+		if err := reserveIps(subnet, ipam); err != nil {
+			klog.Warningf("Failed reserving IPs: %s", err)
+			continue
+		}
+		ipams = append(ipams, ipam)
+	}
+	cache.cache[nodeName] = logicalSwitchInfo{
+		hostSubnets: hostSubnets,
+		ipams:       ipams}
+
+	return nil
+}
+
+func (cache *logicalSwitchCache) Delete(nodeName string) error {
+	cache.Lock()
+	defer cache.Unlock()
+	delete(cache.cache, nodeName)
+	return nil
+}
+
+func (cache *logicalSwitchCache) IsNonHostSubnetSwitch(nodeName string) bool {
+	cache.RLock()
+	defer cache.RUnlock()
+	exists, ok := cache.cache[nodeName]
+	return ok && exists.hostSubnets == nil
+}
+
+func (cache *logicalSwitchCache) GetSwitchSubnet(nodeName string) *net.IPNet {
+	cache.RLock()
+	defer cache.RUnlock()
+	exists, ok := cache.cache[nodeName]
+	if ok && exists.hostSubnets != nil && len(exists.hostSubnets) > 0 {
+		//FIXME DUAL-STACK
+		subnet := *exists.hostSubnets[0]
+		return &subnet
+	}
+	return nil
+}
+
+func (cache *logicalSwitchCache) AllocateIP(nodeName string, ip *net.IP) error {
+	cache.RLock()
+	defer cache.RUnlock()
+	exists, ok := cache.cache[nodeName]
+	// FIXME DUAL-STACK
+	if ok && ip != nil && exists.ipams != nil && len(exists.ipams) > 0 {
+		if err := exists.ipams[0].Allocate(*ip); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (cache *logicalSwitchCache) AllocateNextIP(nodeName string) (*net.IPNet, error) {
+	cache.RLock()
+	defer cache.RUnlock()
+	exists, ok := cache.cache[nodeName]
+	// FIXME DUAL-STACK
+	if ok && len(exists.ipams) == 0 {
+		return nil, fmt.Errorf("failed to allocate IP for node %s because there is no IPAM instance", nodeName)
+	}
+	ip, err := exists.ipams[0].AllocateNext()
+	if err != nil {
+		return nil, err
+	} else if exists.hostSubnets == nil || len(exists.hostSubnets) == 0 {
+		return nil, fmt.Errorf("cannot retrieve subnet mask for node %s during IP allocation", nodeName)
+	}
+	return &net.IPNet{
+		IP:   ip,
+		Mask: exists.hostSubnets[0].Mask,
+	}, nil
+
+}
+
+func (cache *logicalSwitchCache) ReleaseIP(nodeName string, ip *net.IP) error {
+	cache.RLock()
+	defer cache.RUnlock()
+	if ip == nil || nodeName == "" {
+		klog.V(5).Infof("Node name or ip to release is invalid")
+		return nil
+	}
+	exists, ok := cache.cache[nodeName]
+	// FIXME DUAL-STACK
+	if ok && exists.ipams != nil && len(exists.ipams) > 0 {
+		return exists.ipams[0].Release(*ip)
+	}
+	return nil
+}
+
+func reserveIps(subnet *net.IPNet, ipam ipam.Interface) error {
```

Comment:
The naming of this function suggests it can reserve multiple IPs. Why can't this just be logic directly in the allocator, where IPs not available for pods are always skipped?

Reply:
Will move this logic to the allocator once we bring the logic for the ipallocator from k/k to `ovn-org/ovn-kube`

Reply:
done